[Model element] Support `autoplay` in GPU process model element

mwyrzykowski opened the pull request:
#### e77e543069bd20a7fd6585ea9c4792c232715b20
<pre>
[Model element] Support `autoplay` in GPU process model element
<a href="https://bugs.webkit.org/show_bug.cgi?id=299473">https://bugs.webkit.org/show_bug.cgi?id=299473</a>
<a href="https://rdar.apple.com/161268071">rdar://161268071</a>

Reviewed by Sam Weinig.

Add autoplay / pause / play support.

Test: model-element/model-element-animations-playback.html passes all cases
except the one which checks if the USD has an animation as we currently
don&apos;t have the framework support to detect that.

* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::play):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::DDModelPlayer::play):
(WebCore::DDModelPlayer::setAutoplay):
(WebCore::DDModelPlayer::setPaused):
(WebCore::DDModelPlayer::paused const):
* Source/WebGPU/WebGPU/DDMesh.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::DDMesh::play):
(wgpuDDMeshPlay):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(DDBridgeReceiver.setPlaying(_:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::play):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::play):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:

Canonical link: <a href="https://commits.webkit.org/302510@main">https://commits.webkit.org/302510@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06d70cdf72d29f1b8352bea72e24732ae8292e95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136714 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80734 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5fe50c30-89a2-4667-8f3f-2c10c1e6279c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131208 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1470 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98503 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66401 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/739d7add-d598-40e7-af6f-4084684757c8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132284 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/115845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79153 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/59b68caf-8535-40b8-bbc2-7503b4c6c032) 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33974 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79991 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34475 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139187 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1384 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107029 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1426 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112189 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106872 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27208 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1128 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30709 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54020 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1455 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64818 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1272 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1308 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1377 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/Modules/model-element/DDModelPlayer.h:
```diff
@@ -95,6 +100,7 @@ class WEBCORE_EXPORT DDModelPlayer final : public ModelPlayer {
     WeakRef<Page> m_page;
     mutable RefPtr<ModelDisplayBufferDisplayDelegate> m_contentsDisplayDelegate;
     uint32_t m_currentTexture { 0 };
+    std::optional<bool> m_paused;
```
I try to steer people away from std::optional<bool>, as I find it hard to reason about and usually an enum is better suited.

mwyrzykowski replied:
Changed this to an enum

weinig commented on Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
```diff
@@ -143,6 +143,15 @@ void RemoteDDMesh::updateMaterial(const WebCore::DDModel::DDUpdateMaterialDescri
 
+void RemoteDDMesh::play(bool playing)
+{
+#if PLATFORM(COCOA)
```
PLATFORM(COCOA) seems like it's probably the wrong check here. It should be more tailored to what this checking for.

mwyrzykowski replied:
This entire file is in a `GPU_PROCESS_MODEL` check, so this one can be removed

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -1564,4 +1564,11 @@ extension DDBridgeReceiver {
         #endif
     }
+
+    @objc(setPlaying:)
```
is "@objc(setPlaying:)” needed?

mwyrzykowski replied:
Good question, will check before landing

mwyrzykowski replied:
Just the `@objc` part is needed and not the `setPlaying:`, I removed it 👍

mwyrzykowski committed:
[Model element] Support `autoplay` in GPU process model element
https://bugs.webkit.org/show_bug.cgi?id=299473
rdar://161268071

Reviewed by Sam Weinig.

Add autoplay / pause / play support.

Test: model-element/model-element-animations-playback.html passes all cases
except the one which checks if the USD has an animation as we currently
don't have the framework support to detect that.

* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::play):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::DDModelPlayer::play):
(WebCore::DDModelPlayer::setAutoplay):
(WebCore::DDModelPlayer::setPaused):
(WebCore::DDModelPlayer::paused const):
* Source/WebGPU/WebGPU/DDMesh.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::DDMesh::play):
(wgpuDDMeshPlay):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(DDBridgeReceiver.setPlaying(_:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::play):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::play):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:

Canonical link: https://commits.webkit.org/302510@main
